crates/hyperqueue/src/dashboard/ui/screens/home/chart_section.rs

```diff
@@ -0,0 +1,147 @@
+use std::time::{SystemTime, UNIX_EPOCH};
+use tako::messages::gateway::CollectedOverview;
+
+use crate::dashboard::ui::terminal::DashboardFrame;
+use tui::layout::Rect;
+use tui::style::{Color, Modifier, Style};
+use tui::symbols;
+use tui::text::Span;
+use tui::widgets::{Axis, Block, Borders, Chart, Dataset};
+
+/// A new WorkerTimeState is created everytime there is a #worker change event
+#[derive(Clone)]
+struct WorkerStateAtTime {
+    /// The time at which the event being stored happened, wrt. the reference
+    pub at_time: u128,
+    pub event: TimeEventType,
+}
+
+#[derive(Clone)]
+pub enum TimeEventType {
+    /// Current number of workers.
+    WorkerCountUpdate(u32),
+    /// The average cpu utilization of the worker.
+    AvgCpuUsageUpdate(f32),
+}
+
+#[derive(Clone)]
+pub struct ClusterOverviewChart {
+    /// The state of the worker at different times.
+    timed_info_vec: Vec<WorkerStateAtTime>,
+    /// Used to set the min bound of the chart's x-axis
+    time_reference: u128,
+    /// Used to set the max bound of the chart's y-axis
+    max_observed_workers: u64,
+    /// Current number of workers connected
+    current_conn_workers: u64,
+}
+
+impl ClusterOverviewChart {
+    pub fn update(&mut self, overview: CollectedOverview) {
+        // Update the number of workers connected
+        let time = get_current_time();
+        let num_workers = overview.worker_overviews.len();
+        if num_workers > self.max_observed_workers as usize {
+            self.max_observed_workers = num_workers as u64;
+        }
```

Comment:
`self.max_observed_workers = self.max_observed_workers.max(num_workers as u64);`
but I don't think that we have to cache it here.